fs/walk.go

```diff
@@ -250,13 +250,93 @@ func (dt DirTree) Dirs() (dirNames []string) {
 	return dirNames
 }
 
-// Prune remove directories from a directory tree.
-func (dt DirTree) Prune(dirNames []string) {
-	// We do not need to do a recursion.
-	// GC will take care of deleting child entries.
-	for _, dName := range dirNames {
-		delete(dt, dName)
+// Prune remove directories from a directory tree. dirNames contains
+// all directories to remove as keys, with true as values. dirNames
+// will be modified in the function.
+func (dt DirTree) Prune(dirNames map[string]bool) error {
+	// We use map[string]bool to avoid recursion (and potential
+	// stack exhaustion).
+
+	// First we need delete directories from their parents.
+	for dName, remove := range dirNames {
+		if !remove {
+			// Currently all values should be
+			// true, therefore this should not
+			// happen. But this makes function
+			// more predictable.
+			Infof(dName, "Directory in the map for prune, but the value is false")
+			continue
+		}
+		if dName == "" {
+			// if dName is root, do nothing (no parent exist)
+			continue
+		}
+		parent := parentDir(dName)
+		dir, ok := dt[parent]
+		if !ok {
+			return errors.Errorf("parent directory of %s (%s) not in DirTree", dName, parent)
+		}
+		found := false
+		for i, entry := range dir {
+			switch x := entry.(type) {
+			case Directory:
+				if x.Remote() == dName {
```

Comment:
@ncw do we need to take a special care of case-insensitive fs? or in case-insensitive fs naming should be always consistent?